javascript/ql/src/Security/Summaries/ImportFromCsv.qll

```diff
@@ -0,0 +1,192 @@
+  }
+}
+
+/**
+ * An additional sink specified in an `additional-sinks.csv` file.
+ */
+class AdditionalSinkSpec extends ExternalData {
+  AdditionalSinkSpec() {
+    this.getDataPath() = "additional-sinks.csv"
+  }
+
+  /**
+   * Gets the portal specification of this additional sink.
+   */
+  Portal getPortal() {
+    result.toString() = getField(0)
+  }
+
+  /**
+   * Gets the flow label of this sink, or any standard flow label if this sink
+   * does not specify a flow label.
+   */
+  DataFlow::FlowLabel getFlowLabel() {
+    sinkFlowLabelSpec(result, getField(1))
+  }
+
+  /**
+   * Gets the configuration for which this is a sink, or any configuration if
+   * this sink does not specify a configuration.
+   */
+  DataFlow::Configuration getConfiguration() {
+    configSpec(result, getField(2))
+  }
+
+  override string toString() {
+    exists (string labels, string config |
+      labels = strictconcat(getFlowLabel(), " or ") and
+      if getField(2) = "" then
+        config = "any configuration"
+      else
+        config = getConfiguration() |
+      result = getPortal() + " as " + labels + " sink for " + config
+    )
+  }
+}
+
+private class AdditionalSinkFromSpec extends DataFlow::AdditionalSink {
+  AdditionalSinkSpec spec;
+
+  AdditionalSinkFromSpec() {
+    this = spec.getPortal().getAnEntryNode(_)
+  }
+
+  override predicate isSinkFor(DataFlow::Configuration cfg, DataFlow::FlowLabel lbl) {
+    cfg = spec.getConfiguration() and lbl = spec.getFlowLabel()
+  }
+}
+/**
+ * An additional flow step specified in an `additional-steps.csv` file.
+ */
+class AdditionalStepSpec extends ExternalData {
+  AdditionalStepSpec() {
+    this.getDataPath() = "additional-steps.csv"
+  }
+
+  /**
+   * Gets the start portal of this additional step.
+   */
+  Portal getStartPortal() {
+    result.toString() = getField(0)
+  }
+
+  /**
+   * Gets the start flow label of this additional step.
+   */
+  DataFlow::FlowLabel getStartFlowLabel() {
+    result.toString() = getField(1)
+  }
+
+  /**
+   * Gets the end portal of this additional step.
+   */
+  Portal getEndPortal() {
+    result.toString() = getField(2)
+  }
+
+  /**
+   * Gets the end flow label of this additional step.
+   */
+  DataFlow::FlowLabel getEndFlowLabel() {
+    result.toString() = getField(3)
+  }
+
+  /**
+   * Gets the configuration to which this step should be added.
+   */
+  DataFlow::Configuration getConfiguration() {
+    configSpec(result, getField(4))
+  }
+
+  override string toString() {
+    exists (string config |
+      if getField(4) = "" then
+        config = "any configuration"
+      else
+        config = getConfiguration() |
+      result = "edge from " + getStartPortal() + " to " + getEndPortal() +
+               ", transforming " + getStartFlowLabel() + " into " + getEndFlowLabel() +
+               " for " + config
+    )
+  }
+}
+
+private class AdditionalFlowStepFromSpec extends DataFlow::Configuration {
```

Comment:
It would be nice with a dedicated `AdditionalFlowStep` class for this instead, like with the sources and sinks.

Sort of related: The flow summaries only seem to track `TaintTracking::Configuration`. Is that deliberate? Is there a reason we can't use `AdditionalTaintStep`?

Reply:
Hm, good point. There isn't a fundamental reason to only track `TaintTracking::Configuration`s, but it's probably the case we are most interested in for now, so I'll just switch this to `AdditionalTaintStep`.

Reply:
Ugh, except that `AdditionalTaintStep` doesn't currently allow specifying kinds, and it is also `cached` and should only be extended in the standard library. Bit messy, that.

Reply:
Ah, yeah the `cached` thing is a deal-breaker. Still, it seems a bit sneaky to have this configuration-class here, so if we could add an uncached class for this that would be great.